circuit_passes/src/passes/unknown_index_sanitization.rs

```diff
@@ -0,0 +1,328 @@
+            OperatorType::BoolAnd => stack.iter().fold(KnownU32(1), value::bool_and_value),
+            OperatorType::BitOr => resolve_operation(value::bit_or_value, p, &stack),
+            OperatorType::BitAnd => resolve_operation(value::bit_and_value, p, &stack),
+            OperatorType::BitXor => resolve_operation(value::bit_xor_value, p, &stack),
+            OperatorType::PrefixSub => {
+                mod_value(&value::prefix_sub(&stack[0]), p)
+            }
+            OperatorType::BoolNot => KnownU32((!stack[0].to_bool()).into()),
+            OperatorType::Complement => {
+                mod_value(&value::complement(&stack[0]), p)
+            }
+            OperatorType::ToAddress => value::to_address(&stack[0]),
+            OperatorType::MulAddress => stack.iter().fold(KnownU32(1), value::mul_address),
+            OperatorType::AddAddress => stack.iter().fold(KnownU32(0), value::add_address),
+        });
+        (computed_value, env)
+    }
+
+    pub fn execute_instruction<'env>(&self, inst: &'env InstructionPointer, env: Env<'env>) -> R<'env> {
+        match inst.as_ref() {
+            Instruction::Value(b) => self.execute_value_bucket(b, env),
+            Instruction::Load(b) => self.execute_load_bucket(b, env),
+            Instruction::Compute(b) => self.execute_compute_bucket(b, env),
+            _ => unreachable!(),
+        }
+    }
+}
+
+pub struct UnknownIndexSanitizationPass {
+    // Wrapped in a RefCell because the reference to the static analysis is immutable but we need mutability
+    memory: RefCell<PassMemory>,
+    load_replacements: RefCell<BTreeMap<LoadBucket, Range<usize>>>,
+    store_replacements: RefCell<BTreeMap<StoreBucket, Range<usize>>>,
+}
+
+
+/**
+ * The goal of this pass is to
+ */
+impl UnknownIndexSanitizationPass {
+    pub fn new(prime: &String) -> Self {
+        UnknownIndexSanitizationPass {
+            memory: PassMemory::new_cell(prime, "".to_string(), Default::default()),
+            load_replacements: Default::default(),
+            store_replacements: Default::default(),
+        }
+    }
+
+    fn find_bounds(&self, address: &AddressType, location: &LocationRule, env: &Env) -> Range<usize> {
+        let mem = self.memory.borrow();
+        let interpreter = ZeroingInterpreter::init(&mem.prime, &mem.constant_fields);
+        let current_scope = &mem.current_scope;
+
+        let mapping = match address {
+            AddressType::Variable => &mem.variables_index_mapping[current_scope],
+            AddressType::Signal => &mem.signal_index_mapping[current_scope],
+            AddressType::SubcmpSignal { cmp_address, .. } => &mem.component_addr_index_mapping[current_scope],
+        };
+
+        /*
+         * We assume locations are of the form:
+         *      (base_offset + (mul_offset * UNKNOWN))
+         * So, if we set the unknown value to 0, we will compute the base offset,
+         * which will let us look up the range of the underlying array.
+         */
```

Comment:
In this comment add that that form should be for 1d arrays but it expands for multiple dimensions following an similar approach

Reply:
Addressed